### PR TITLE
Refactor record sort to ensure position is maintained

### PR DESCRIPTION
**Pull request by kylebaron**
Issue reported in #46 . 

This refactor makes use of the `pos` attribute to track the position of each record in the data set. When additional doses are in the mix, pos can either take a large value (bigger than the number of rows in the data set) or a small value (negative 1). When `pos` is large, additional doses scheduled at the same time as records in the data set will be pushed forward (toward the end of the data set) so that observations at the same time are at the end of the dosing interval. In other words, the observations, while having the same time, will happen first (in order) so TAD will be 24 for once-daily dosing rather than 0. This is the default behavior. When `pos` is smaller than all the other records in the data set, then the addl dose record will happen first and observations at the same time will be viewed as happening at that same time (TAD = 0).

### CODE REVIEW

File: src/lastdose.cpp [hunk 1]
```diff
@@ -38,13 +38,10 @@ bool is_dose(const int evid, const bool comment) {
 typedef std::vector<record> recs;
 
 bool Comp1(const record& a, const record& b) {
```
Comment (kylebaron) [inline src/lastdose.cpp:40]: There is only one comparison function required now to sort the records.

File: src/lastdose.cpp [hunk 2]
```diff
@@ -60,13 +57,14 @@ Rcpp::List lastdose_impl(Rcpp::NumericVector id,
                          Rcpp::LogicalVector comment) {
 
   amt = Rcpp::clone(amt);
-  bool use_comp1 = sort1[0];
+  bool obs_first = sort1[0];
```
Comment (kylebaron) [inline src/lastdose.cpp:60]: Renaming this for transparency.

File: src/lastdose.cpp [hunk 3]
```diff
@@ -60,13 +57,14 @@ Rcpp::List lastdose_impl(Rcpp::NumericVector id,
...
   bool use_fill = !back_calc[0];
   std::vector<double> idn;
   std::vector<int> idstart;
   std::vector<int> idend;
   double lastid = -1E9;
   int nrows = id.size();
+  const int max_addl_pos = nrows + 1;
```
Comment (kylebaron) [inline src/lastdose.cpp]: When `obs_first` is `TRUE`, then any record from additional doses will always get pushed toward the bottom of the data set at a given time when sorted.

File: src/lastdose.cpp [hunk 4]
```diff
@@ -152,18 +150,15 @@ Rcpp::List lastdose_impl(Rcpp::NumericVector id,
         }
         for(int k = 0; k < addl[j]; ++k) {
           record addl_rec(0.0,amt[j],evid[j],false,false);
+          addl_rec.pos = obs_first ? max_addl_pos : -1;
```
Comment (kylebaron) [inline src/lastdose.cpp:153]: `pos` attribute will be used to sort records.

File: src/lastdose.cpp [hunk 5]
```diff
@@ -152,18 +150,15 @@ Rcpp::List lastdose_impl(Rcpp::NumericVector id,
...
           addl_rec.time = time[j] + ii[j]*double(k+1);
           if(addl_rec.time >= (max_time)) break;
           this_id.push_back(addl_rec);
         }
       }
       ++crow;
     }
-    if(use_comp1) {
```
Comment (kylebaron) [inline src/lastdose.cpp:162]: Only one comparison function is needed now.

**Comment by kylebaron** [comment]
It looks like tests are failing on R 3.6. I'm not sure why that might be ... those same tests were passing at previous release and I don't _think_ that code was changed at all. Seems like a CI issue @kyleam

**Comment by kyleam** [comment]
> those same tests were passing at previous release and I don't _think_ that code was changed at all

Hmm, the failure is at test-lastdose.R line 385.  Isn't that the test being added in this PR?

```
══ Warnings ════════════════════════════════════════════════════════════════════
1. comments vector is subset for NA time #38 [LSD-TEST-025] ('test-lastdose.R:359:3') - looking for comment records; found column `C` but it wasn't character or logical

══ Failed ══════════════════════════════════════════════════════════════════════
── 1. Error ('test-lastdose.R:385:3'): TAD is the same for records with the same
Error: id column is required to be numeric
Backtrace:
    ▆
 1. └─lastdose::lastdose(data, addl_ties = "obs_first") at test-lastdose.R:385:3
 2.   └─lastdose::lastdose_list(data, ...)
```

(Not for this PR, but it would be good to move lastdose from Drone to GitHub Actions.  It's one of the few packages that haven't been transitioned.  I can help with that before the next release.)

**Comment by kyleam** [comment]
I pushed a scratch branch with `main`'s content and confirmed that 3.6 passes: https://github-drone.metrumrg.com/metrumresearchgroup/lastdose/190

File: src/lastdose.cpp (hunk 1 above)
Comment (andersone1) [reply src/lastdose.cpp:40]: looks good

File: src/lastdose.cpp (hunk 2 above)
Comment (andersone1) [reply src/lastdose.cpp:60]: looks good

**Review by andersone1 — APPROVED** [review]

**Comment by andersone1** [comment]
@kylebaron 

Changes look good, I approved the PR in it's current state.

If another commit is necessary for it to pass the CI lmk, and I can review that as well.

**Comment by kyleam** [comment]
@kylebaron R 3.6 runs are green after test change in ba0bbe2

https://github-drone.metrumrg.com/metrumresearchgroup/lastdose/192
https://github-drone.metrumrg.com/metrumresearchgroup/lastdose/191

**Comment by kyleam** [comment]
I said:

> it would be good to move lastdose from Drone to GitHub Actions. It's one of the few packages that haven't been transitioned. I can help with that before the next release.

PR for that is at gh-48.

**Comment by kylebaron** [comment]
Thanks for your help tracking that down, @kyleam .  I thought the trouble was coming from a different part of the code; but yeah this is obviously as you diagnosed.

**Pull request merged**